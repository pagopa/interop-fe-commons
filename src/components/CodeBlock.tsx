import React from 'react'
import { Box, Paper, type BoxProps } from '@mui/material'
import { CopyToClipboardButton } from '@pagopa/mui-italia'

interface CodeBlockProps extends BoxProps {
  /**
   * The code to display
   * If the code is an object, it will be stringified with 2 spaces indentation
   */
  code: unknown
  /**
   * If true, the copy button will be hidden
   * @default false
   * */
  hideCopyButton?: boolean
}

export const CodeBlock: React.FC<CodeBlockProps> = ({ code, hideCopyButton, ...boxProps }) => {
  const stringifiedCode = typeof code === 'object' ? JSON.stringify(code, null, 2) : String(code)
  return (
    <Box {...boxProps}>
      <Paper
        sx={{
          p: 2,
          position: 'relative',
          backgroundColor: 'background.default',
          whiteSpace: 'pre-wrap',
          wordBreak: 'break-word',
          maxHeight: 520,
          overflowY: 'auto',
          overflowX: 'hidden',
          fontSize: 'small',
        }}
      >
        {!hideCopyButton && (
          <Box sx={{ position: 'absolute', top: 8, right: 8 }}>
            <CopyToClipboardButton value={stringifiedCode} />
          </Box>
        )}
        <code>{stringifiedCode}</code>
      </Paper>
    </Box>
  )
}
