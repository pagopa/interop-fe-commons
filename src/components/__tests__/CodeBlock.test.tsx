import React from 'react'
import { CodeBlock } from '../CodeBlock'
import renderer from 'react-test-renderer'

const testCode = {
  test: 'test',
}

describe('CodeBlock component', () => {
  it('Renders correctly', () => {
    const tree = renderer.create(<CodeBlock code={testCode} />).toJSON()
    expect(tree).toMatchSnapshot()
  })

  it('Render correctly without copy to clipboard button', () => {
    const tree = renderer.create(<CodeBlock code={testCode} hideCopyButton />).toJSON()
    expect(tree).toMatchSnapshot()
  })
})
