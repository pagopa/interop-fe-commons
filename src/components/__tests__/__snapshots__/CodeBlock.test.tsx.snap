// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`CodeBlock component > Render correctly without copy to clipboard button 1`] = `
<div
  className="MuiBox-root css-0"
>
  <div
    className="MuiPaper-root MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation1 css-kmr1o1-MuiPaper-root"
  >
    <code>
      {
  "test": "test"
}
    </code>
  </div>
</div>
`;

exports[`CodeBlock component > Renders correctly 1`] = `
<div
  className="MuiBox-root css-0"
>
  <div
    className="MuiPaper-root MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation1 css-kmr1o1-MuiPaper-root"
  >
    <div
      className="MuiBox-root css-1q9qc6z"
    >
      <button
        aria-label="Copia"
        aria-labelledby={null}
        className="MuiButtonBase-root MuiIconButton-root MuiIconButton-sizeMedium css-vqczhe-MuiButtonBase-root-MuiIconButton-root"
        data-mui-internal-clone-element={true}
        disabled={false}
        onBlur={[Function]}
        onClick={[Function]}
        onContextMenu={[Function]}
        onDragLeave={[Function]}
        onFocus={[Function]}
        onKeyDown={[Function]}
        onKeyUp={[Function]}
        onMouseDown={[Function]}
        onMouseLeave={[Function]}
        onMouseOver={[Function]}
        onMouseUp={[Function]}
        onTouchEnd={[Function]}
        onTouchMove={[Function]}
        onTouchStart={[Function]}
        role="button"
        tabIndex={0}
        type="button"
      >
        <svg
          aria-hidden={true}
          className="MuiSvgIcon-root MuiSvgIcon-fontSizeSmall css-ptiqhd-MuiSvgIcon-root"
          data-testid="ContentCopyIcon"
          focusable="false"
          viewBox="0 0 24 24"
        >
          <path
            d="M16 1H4c-1.1 0-2 .9-2 2v14h2V3h12V1zm3 4H8c-1.1 0-2 .9-2 2v14c0 1.1.9 2 2 2h11c1.1 0 2-.9 2-2V7c0-1.1-.9-2-2-2zm0 16H8V7h11v14z"
          />
        </svg>
        <span
          className="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
        />
      </button>
    </div>
    <code>
      {
  "test": "test"
}
    </code>
  </div>
</div>
`;
